src/number_theory/modular.lean

```diff
@@ -371,6 +400,266 @@ begin
+      by nlinarith
+    ... = norm_sq (denom g z) : by simp [norm_sq]; ring,
+  have zIm : (3 : ℝ) / 4 < (z.im)^2 := ImGeInD _ hz,
+  /- This is the main calculation:
+  `sqrt 3 / 2 < Im(g•z) = Im(z)/|cz+d|^2 ≤ y/(c^2 y^2) < 2/(c^2 sqrt 3)`
+  -/
+  calc
+  (3 : ℝ) / 4 < ((g • z).im) ^ 2 : ImGeInD _ hg
+  ... = (z.im) ^ 2 / (norm_sq (denom g z)) ^ 2 : _
+  ... ≤ (1 : ℝ) / ((↑ₘg 1 0) ^ 4 * (z.im) ^ 2) : _
+  ... < (4 : ℝ) / (3 * (↑ₘg 1 0) ^ 4) : _,
+  { convert congr_arg (λ (x:ℝ), x ^ 2) (im_smul_eq_div_norm_sq g z) using 1,
+    exact (div_pow _ _ 2).symm, },
+  { rw div_le_div_iff,
+    convert pow_le_pow_of_le_left _ czPdGecy 2 using 1;
+    ring_nf,
+    { nlinarith, },
+    { exact pow_two_pos_of_ne_zero _ (norm_sq_denom_ne_zero g z), },
+    { nlinarith, }, },
+  { rw div_lt_div_iff,
+    repeat {nlinarith}, },
+end
+
+/-- Knowing that `3/4 < 4/(3c^4)` from `ineq_1`, and `c≠0`, we conclude that `c = ±1`. -/
+lemma ineq_2 (c : ℤ) (hc₁ : (3 : ℝ)/4 < 4/ (3* c^4)) (hc₂ : c ≠ 0) : c = 1 ∨ c = -1 :=
+begin
+  rcases le_or_gt (|c|) 1 with h | (h : 2 ≤ |c|),
+  { -- case |c| ≤ 1
+    obtain ⟨h1c, hc1⟩ : -1 ≤ c ∧ c ≤ 1 := abs_le.mp h,
+    interval_cases c; tauto },
+  { -- case 2 ≤ |c|
+    exfalso,
+    have : 2^4 ≤ c^4,
+    { refine pow_four_le_pow_four _,
+      convert h using 1, },
+    have : (2:ℝ)^4 ≤ c^4,
+    { norm_cast,
+      convert this using 1, },
+    have := (div_lt_div_iff _ _).mp hc₁,
+    repeat {linarith}, },
+end
+
+lemma coe_T : ↑ₘT = ![![1, 1], ![0, 1]] := rfl
+
+lemma coe_T_inv : ↑ₘ(T⁻¹) = ![![1, -1], ![0, 1]] :=
+begin
+  rw [coe_inv, coe_T, adjugate_fin_two],
+  simp,
+end
+
+/-- `coe_T_zpow` is the matrix `T` raised to the power `n : ℤ`. -/
+lemma coe_T_zpow (n : ℤ) : ↑ₘ(T ^ n) = ![![1, n], ![0,1]] :=
+begin
+  induction n using int.induction_on with n h n h,
+  { rw [zpow_zero, coe_one],  ext i j,
+    fin_cases i; fin_cases j; simp, },
+  { rw [zpow_add, zpow_one, coe_mul, h, coe_T], ext i j,
+    fin_cases i; fin_cases j; simp [matrix.mul, dot_product, fin.sum_univ_succ]; ring, },
+  { rw [zpow_sub, zpow_one, coe_mul, h, coe_T_inv], ext i j,
+    fin_cases i; fin_cases j; simp [matrix.mul, dot_product, fin.sum_univ_succ]; ring, },
+end
+
+/- If `c = 1`, then `g = [[1,a],[0,1]] * S * [[1,d],[0,1]]`. -/
+lemma g_eq_of_c_eq_one (g : SL(2,ℤ)) (hc : ↑ₘg 1 0 = 1) :
+  g = T^(↑ₘg 0 0) * S * T^(↑ₘg 1 1) :=
+begin
+  ext i j, fin_cases i; fin_cases j,
+  { simp [S, coe_T_zpow, matrix.mul_apply, fin.sum_univ_succ] },
+  { have g_det : (1:ℤ) = ↑ₘg 0 0 * ↑ₘg 1 1 - 1 * ↑ₘg 0 1,
+    { convert det_fin_two ↑ₘg using 1,
+      { rw g.det_coe },
+      rw hc,
+      ring },
+    simp [S, coe_T_zpow, matrix.mul_apply, fin.sum_univ_succ],
+    rw g_det,
+    simp, },
+  { simpa [S, coe_T_zpow, matrix.mul_apply, fin.sum_univ_succ] using hc },
+  { simp [S, coe_T_zpow, matrix.mul_apply, fin.sum_univ_succ], },
+end
+
+lemma cast_one_le_of_pos {α : Type*} [linear_ordered_ring α] {n : ℤ} (hn : 0 < n) :
```

Comment:
I think the lemmas `cast_one_le_of_pos`, `cast_le_neg_one_of_neg`, `nneg_mul_add_sq_of_abs_le_one` could all go in `data/int/cast.lean` about here: 
https://github.com/leanprover-community/mathlib/blob/49e41eb5383e372352490c6236cb0293e4e84d59/src/data/int/cast.lean#L191

This will also place them into the `int` namespace which is appropriate.